ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/ChainHeadPrivateNonceProvider.java

```diff
@@ -21,35 +21,73 @@
+import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
 
 public class ChainHeadPrivateNonceProvider implements PrivateNonceProvider {
+  private static final Logger LOG = LogManager.getLogger();
   private final Blockchain blockchain;
   private final PrivateStateRootResolver privateStateRootResolver;
   private final WorldStateArchive privateWorldStateArchive;
+  private final PmtTransactionPool pmtTransactionPool;
 
   public ChainHeadPrivateNonceProvider(
       final Blockchain blockchain,
       final PrivateStateRootResolver privateStateRootResolver,
-      final WorldStateArchive privateWorldStateArchive) {
+      final WorldStateArchive privateWorldStateArchive,
+      final PmtTransactionPool pmtTransactionPool) {
     this.blockchain = blockchain;
     this.privateStateRootResolver = privateStateRootResolver;
     this.privateWorldStateArchive = privateWorldStateArchive;
+    this.pmtTransactionPool = pmtTransactionPool;
   }
 
+  /**
+   * Calculate the nonce, while taking into account any PMTs that are already in progress. This
+   * makes nonce management for private tx slightly more robust.
+   *
+   * @param sender the sender of the transaction
+   * @param privacyGroupId the privacy group ID this tx is for
+   * @return the nonce, taking into account the PMT in the pool
+   */
   @Override
   public long getNonce(final Address sender, final Bytes32 privacyGroupId) {
     final BlockHeader chainHeadHeader = blockchain.getChainHeadHeader();
-    Hash chainHeadHash = chainHeadHeader.getHash();
+    final Hash chainHeadHash = chainHeadHeader.getHash();
+
+    LOG.info(
+        "checking for PMT matches for sender "
+            + sender
+            + " and privacyGroupID (base 64) "
+            + privacyGroupId.toBase64String());
+    // if latestNonce > pendingNonce return latestNonce (as below)
+    // else return pendingNonce + 1
+    long pendingPrivateNonceFromPmtPool =
+        pmtTransactionPool.getMaxMatchingNonce(
+            sender.toHexString(), privacyGroupId.toBase64String());
+
     final Hash stateRoot =
         privateStateRootResolver.resolveLastStateRoot(privacyGroupId, chainHeadHash);
-    return privateWorldStateArchive
-        .get(stateRoot, chainHeadHash)
-        .map(
-            privateWorldState -> {
-              final Account account = privateWorldState.get(sender);
-              return account == null ? 0L : account.getNonce();
-            })
-        .orElse(Account.DEFAULT_NONCE);
+    long stateBasedPrivateNonce =
+        privateWorldStateArchive
+            .get(stateRoot, chainHeadHash)
+            .map(
+                privateWorldState -> {
+                  final Account account = privateWorldState.get(sender);
+                  return account == null ? 0L : account.getNonce();
+                })
+            .orElse(Account.DEFAULT_NONCE);
+
+    if (pendingPrivateNonceFromPmtPool == 0 && stateBasedPrivateNonce == 0) {
```

Comment:
Isn't this case redundant and covered by the ones below?

Reply:
gone